src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/InstrumentRecorder.cs

```diff
@@ -0,0 +1,184 @@
+        public InstrumentRecorder(object? scopeFilter, string meterName, string instrumentName)
+        {
+            if (meterName is null)
+            {
+                throw new ArgumentNullException(nameof(meterName));
+            }
+
+            if (instrumentName is null)
+            {
+                throw new ArgumentNullException(nameof(instrumentName));
+            }
+
+            _measurements = new List<Measurement<T>>();
+
+            _meterListener = new MeterListener();
+            _meterListener.InstrumentPublished = (instrument, listener) =>
+            {
+                if (object.Equals(instrument.Meter.Scope, scopeFilter) &&
+                    instrument.Meter.Name == meterName &&
+                    instrument.Name == instrumentName &&
+                    (instrument is ObservableInstrument<T> || instrument is Instrument<T>))
+                {
+                    if (instrument is ObservableInstrument<T>)
+                    {
+                        _isObservableInstrument = true;
+                    }
+
+                    listener.EnableMeasurementEvents(instrument, state: null);
+                    Instrument = instrument;
+                }
+            };
+
+            _meterListener.SetMeasurementEventCallback<T>(OnMeasurementRecorded);
+            _meterListener.Start();
+        }
+
+        /// <summary>
+        /// Initialize a new instance <see cref="InstrumentRecorder{T}" /> to record the measurement published by an <see cref="Instrument" /> with the name <paramref name="instrumentName" />
+        /// and the meter <paramref name="meter"/>.
+        /// </summary>
+        /// <param name="meter">The meter that produced the instrument required for recording the published measurements.</param>
+        /// <param name="instrumentName">The name of the instrument.</param>
+        public InstrumentRecorder(Meter meter, string instrumentName)
+        {
+            if (meter is null)
+            {
+                throw new ArgumentNullException(nameof(meter));
+            }
+
+            if (instrumentName is null)
+            {
+                throw new ArgumentNullException(nameof(instrumentName));
+            }
+
+            _measurements = new List<Measurement<T>>();
+
+            _meterListener = new MeterListener();
+            _meterListener.InstrumentPublished = (instrument, listener) =>
+            {
+                if (object.ReferenceEquals(instrument.Meter, meter) && instrument.Name == instrumentName &&
+                    (instrument is ObservableInstrument<T> || instrument is Instrument<T>))
+                {
+                    if (instrument is ObservableInstrument<T>)
+                    {
+                        _isObservableInstrument = true;
+                    }
+
+                    listener.EnableMeasurementEvents(instrument, state: null);
+                    Instrument = instrument;
+                }
+            };
```

Comment:
Can this callback be invoked multiple times concurrently?

Reply:
No. The calls to the listener are serialized.

Reply:
After chatting offline we addressed this by using interlocked compare exchange.